maas-agent/src/charm.py

```diff
@@ -139,6 +140,21 @@ def _on_remove(self, _event: ops.RemoveEvent) -> None:
         except Exception as ex:
             logger.error(str(ex))
 
+    def _on_upgrade(self, _event: ops.UpgradeCharmEvent) -> None:
+        """Upgrade MAAS install on the machine.
+
+        Args:
+            event (ops.UpgradeCharmEvent): Event from ops framework
+        """
+        self.unit.status = ops.MaintenanceStatus("upgrading...")
+        channel = str(self.config.get("channel", MAAS_SNAP_CHANNEL))
+        try:
```

Comment:
1) We do not want to run downgrades. If we detect such behavior we need to error out
2) We do not want to run refresh on same channel e.g. `3.4/stable` -> `3.4/stable` since it may end up with maas-region or maas-agent units on different revisions as others (e.g., we had maas 3.4.3 and the refresh would bring 3.4.4). So if we detect same channel, we do nothing (the current behavior)
3) if refresh is what should happen e.g., `3.4/stable` -> `3.5/stable` we run the bits of codes you added with your PR.